tensorflow/c/eager/dlpack.cc

```diff
@@ -0,0 +1,292 @@
+        default:
+          status->status = tensorflow::errors::InvalidArgument(
+              "Unsupported UInt bits: ", dtype.bits);
+      }
+      break;
+    case DLDataTypeCode::kDLInt:
+      switch (dtype.bits) {
+        case 8:
+          tf_dtype = TF_DataType::TF_INT8;
+          break;
+        case 16:
+          tf_dtype = TF_DataType::TF_INT16;
+          break;
+        case 32:
+          tf_dtype = TF_DataType::TF_INT32;
+          break;
+        case 64:
+          tf_dtype = TF_DataType::TF_INT64;
+          break;
+        default:
+          status->status = tensorflow::errors::InvalidArgument(
+              "Unsupported Int bits: ", dtype.bits);
+      }
+      break;
+    case DLDataTypeCode::kDLFloat:
+      switch (dtype.bits) {
+        case 16:
+          tf_dtype = TF_DataType::TF_HALF;
+          break;
+        case 32:
+          tf_dtype = TF_DataType::TF_FLOAT;
+          break;
+        case 64:
+          tf_dtype = TF_DataType::TF_DOUBLE;
+          break;
+        default:
+          status->status = tensorflow::errors::InvalidArgument(
+              "Unsupported Float bits: ", dtype.bits);
+      }
+      break;
+    case DLDataTypeCode::kDLBfloat:
+      switch (dtype.bits) {
+        case 16:
+          tf_dtype = TF_DataType::TF_BFLOAT16;
+          break;
+        default:
+          status->status = tensorflow::errors::InvalidArgument(
+              "Unsupported BFloat bits: ", dtype.bits);
+      }
+      break;
+    default:
+      status->status = tensorflow::errors::InvalidArgument(
+          "Unsupported Type Codes: ", dtype.code);
+  }
+
+  return tf_dtype;
+}
+
+void DeallocatorWrapperFunc(void* data, size_t len, void* dlmt_vptr) {
+  DLManagedTensor* dlmt = static_cast<DLManagedTensor*>(dlmt_vptr);
+  dlmt->deleter(const_cast<DLManagedTensor*>(dlmt));
+}
+
+}  // namespace
+
+void TFE_CallDLManagedTensorDeleter(void* dlm_ptr) {
+  DLManagedTensor* dlMTensor = static_cast<DLManagedTensor*>(dlm_ptr);
+  if (dlMTensor) {
+    dlMTensor->deleter(const_cast<DLManagedTensor*>(dlMTensor));
+  }
+}
+
+void* TFE_HandleToDLPack(TFE_TensorHandle* h, TF_Status* status) {
+  DLManagedTensor* tfdlmtensor = TFEHandleToTFDLManagedTensorCtx(h, status);
+  return static_cast<void*>(tfdlmtensor);
+}
+
+TFE_TensorHandle* TFE_HandleFromDLPack(void* dlm, TF_Status* status) {
+  TFE_ContextOptions* opts = TFE_NewContextOptions();
+  TFE_Context* ctx = TFE_NewContext(opts, status);
+  DLManagedTensor* dlmt = static_cast<DLManagedTensor*>(dlm);
+
+  absl::optional<std::string> device_name =
+      DeviceNameFromDlContext(dlmt->dl_tensor.ctx, status);
+  if (!device_name.has_value()) {
+    status->status =
+        tensorflow::errors::InvalidArgument("Unsupported Device Type");
+    return nullptr;
+  }
+  TF_DataType dtype = TfDataTypeFormDlDataType(dlmt->dl_tensor.dtype, status);
+  int num_dims = dlmt->dl_tensor.ndim;
+  const int64_t* dims = dlmt->dl_tensor.shape;
+  void* data = dlmt->dl_tensor.data;
+
```

Comment:
Don't you need to check that the strides are major-to-minor here if not `nullptr`?

Reply:
According to @sanjoy there's no strides in `tensorflow::Tensor`? I'm not familiar with TF, could you give more suggestion on this?

Reply:
Here you are importing a DLPack tensor that may have non-trivial strides, even though TF does not support them. So you must do something, and I think the only reasonable thing to do is to detect that case and fail. (i.e., where the strides in the DLPack tensor do not imply the canonical major-to-minor order of elements that TF expects.)

Reply:
Check added